src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1417,58 +1420,36 @@ uct_dc_mlx5_iface_dci_do_rand_pending_tx(ucs_arbiter_t *arbiter,
     return res;
 }
 
-ucs_arbiter_cb_result_t
-uct_dc_mlx5_ep_arbiter_purge_internal_cb(ucs_arbiter_t *arbiter,
-                                         ucs_arbiter_group_t *group,
-                                         ucs_arbiter_elem_t *elem, void *arg)
+static ucs_arbiter_cb_result_t
+uct_dc_mlx5_ep_arbiter_purge_cb(ucs_arbiter_t *arbiter, ucs_arbiter_group_t *group,
+                                ucs_arbiter_elem_t *elem, void *arg)
 {
-    uct_dc_mlx5_ep_t *ep         = arg;
+    uct_purge_cb_args_t *cb_args = arg;
+    void **priv_args             = cb_args->arg;
+    uct_dc_mlx5_ep_t *ep         = priv_args[0];
     uct_dc_mlx5_iface_t *iface   = ucs_derived_of(ep->super.super.iface,
                                                   uct_dc_mlx5_iface_t);
     uct_pending_req_t *req       = ucs_container_of(elem, uct_pending_req_t,
                                                     priv);
-    uct_rc_pending_req_t *freq;
 
     if (uct_dc_mlx5_iface_is_dci_rand(iface) &&
         (uct_dc_mlx5_pending_req_priv(req)->ep != ep)) {
-        /* Element belongs to another ep - do not remove it */
+        /* Element belongs to another endpoint - do not remove it */
         return UCS_ARBITER_CB_RESULT_NEXT_GROUP;
     }
 
-    if (ucs_unlikely(req->func == uct_dc_mlx5_iface_fc_grant)) {
-        /* User callback should not be called for FC messages. Just return
-         * pending request memory to the pool */
-        freq = ucs_derived_of(req, uct_rc_pending_req_t);
-        ucs_mpool_put(freq);
-        return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
-    }
+    ucs_assertv((ep != iface->tx.fc_ep) ||
```

Comment:
how could it be fc_ep here?

Reply:
it is called from FC enpoint cleanup:
https://github.com/openucx/ucx/blob/f4efac3842ea8b9ef56d0b2be907bb04ab95d91f/src/uct/ib/dc/dc_mlx5.c#L1096

Reply:
seems to me this PR replaced that code by `uct_dc_mlx5_iface_fc_ep_pending_purge_cb`

Reply:
> seems to me this PR replaced that code by uct_dc_mlx5_iface_fc_ep_pending_purge_cb

`uct_dc_mlx5_ep_pending_purge` passes `uct_dc_mlx5_ep_arbiter_purge_cb` to `ucs_arbiter_dispatch` and then `uct_dc_mlx5_ep_arbiter_purge_cb` invokes `uct_dc_mlx5_iface_fc_ep_pending_purge_cb`

Reply:
probably missing something, why not have separate flows for purging fc_ep and regular dc ep? IMO they don't have anything relevant in common